asset-registry/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+use xcm::VersionedMultiLocation;
+
+pub use impls::*;
+pub use weights::WeightInfo;
+
+mod impls;
+mod mock;
+mod tests;
+mod weights;
+
+#[derive(scale_info::TypeInfo, Encode, Decode, Clone, Eq, PartialEq, Debug)]
+pub struct AssetMetadata<Balance, CustomMetadata: Parameter + Member + TypeInfo> {
+	pub decimals: u32,
+	pub name: Vec<u8>,
+	pub symbol: Vec<u8>,
+	pub existential_deposit: Balance,
+	pub location: Option<VersionedMultiLocation>,
+	pub additional: CustomMetadata,
+}
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Additional non-standard metadata to store for each asset
+		type CustomMetadata: Parameter + Member + TypeInfo;
+
+		/// The type used as a unique asset id,
+		type AssetId: Parameter + Member + TypeInfo;
+
+		/// The origin that is allowed to manipulate metadata.
+		type AuthorityOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+
+		/// A filter ran upon metadata registration that assigns an is and
+		/// potentially modifies the supplied metadata.
+		type ProcessAsset: AssetProcessor<Self::AssetId, AssetMetadata<Self::Balance, Self::CustomMetadata>>;
+
+		/// The balance type.
+		type Balance: Parameter
+			+ Member
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Into<u128>;
+
+		// /// Weight information for extrinsics in this module.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Asset was not found
+		AssetNotFound,
+		/// The version of the `Versioned` value used is not able to be
+		/// interpreted.
+		BadVersion,
+		/// The asset id is invalid.
+		InvalidAssetId,
+		/// Another asset was already register with this location.
+		ConflictingLocation,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RegisteredAsset {
+			asset_id: T::AssetId,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+		},
+		UpdatedAsset {
+			asset_id: T::AssetId,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+		},
+		SetLocation {
+			asset_id: T::AssetId,
+			location: VersionedMultiLocation,
+		},
+	}
+
+	/// The total issuance of a token type.
+	#[pallet::storage]
+	#[pallet::getter(fn get_metadata)]
+	pub type Metadata<T: Config> =
+		StorageMap<_, Twox64Concat, T::AssetId, AssetMetadata<T::Balance, T::CustomMetadata>, OptionQuery>;
+
+	/// The total issuance of a token type.
+	#[pallet::storage]
+	#[pallet::getter(fn get_asset_id)]
+	pub type MultiLocationLookup<T: Config> = StorageMap<_, Twox64Concat, MultiLocation, T::AssetId, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn last_asset_id)]
+	pub(crate) type LastAssetId<T: Config> = StorageValue<_, T::AssetId, OptionQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		_phantom: PhantomData<T>,
```

Comment:
should have a way to register assets in genesis, but it can happen in later PR

Reply:
Hmm yea I actually just ran into a problem when trying to implement that: it requires the `AssetMetadata` to implement Serialize & Deserialize, which `VersionedMultiLocation` does not implement.. We'd have to write a custom implementation, which is certainly possible, but I'd leave it for a future PR indeed